contrib/gitannex/gitannex.go

```diff
@@ -0,0 +1,447 @@
+	return []configDefinition{
+		{
+			"rcloneremotename",
+			"Name of the rclone remote to use. " +
+				"Must match a remote known to rclone. " +
+				"(Note that rclone remotes are a distinct concept from git-annex remotes.)",
+			&s.configRcloneRemoteName,
+		},
+		{
+			"rcloneprefix",
+			"Directory where rclone will write git-annex content. " +
+				"If not specified, defaults to \"git-annex-rclone\". " +
+				"This directory be created on init if it does not exist.",
+			&s.configPrefix,
+		},
+	}
+}
+
+// Query git-annex for config values.
+func (s *server) queryConfigs() error {
+	if s.configsDone {
+		return nil
+	}
+
+	// Send a "GETCONFIG" message for each required config and parse git-annex's
+	// "VALUE" response.
+	for _, config := range s.getRequiredConfigs() {
+		s.sendMsg(fmt.Sprintf("GETCONFIG %s", config.name))
+
+		msg, err := s.getMsg()
+		if err != nil {
+			return err
+		}
+		valueKeyword, value, wasFound := strings.Cut(msg, " ")
+		if !wasFound || valueKeyword != "VALUE" {
+			return fmt.Errorf("failed to parse message: %s", msg)
+		}
+		value = strings.TrimSpace(value)
+		if value == "" {
+			return fmt.Errorf("config value of %q must not be empty", config.name)
+		}
+
+		*config.destination = value
+	}
+
+	s.configsDone = true
+	return nil
+}
+
+func (s *server) handlePrepare() error {
+	if err := s.queryConfigs(); err != nil {
+		s.sendMsg("PREPARE-FAILURE Error getting configs")
+		return fmt.Errorf("error getting configs: %w", err)
+	}
+	s.sendMsg("PREPARE-SUCCESS")
+	return nil
+}
+
+// Git-annex is asking us to return the list of settings that we use. Keep this
+// in sync with `handlePrepare()`.
+func (s *server) handleListConfigs() {
+	for _, config := range s.getRequiredConfigs() {
+		s.sendMsg(fmt.Sprintf("CONFIG %s %s", config.name, config.description))
+	}
+	s.sendMsg("CONFIGEND")
+}
+
+func (s *server) handleTransfer(pieces []string) error {
+	if len(pieces) != 3 {
+		s.sendMsg("ERROR malformed arguments for TRANSFER")
+		return fmt.Errorf("malformed arguments for TRANSFER: %v", pieces)
+	}
+
+	argMode, argKey, argFile := pieces[0], pieces[1], pieces[2]
+
+	if err := s.queryConfigs(); err != nil {
+		s.sendMsg(fmt.Sprintf("TRANSFER-FAILURE %s %s failed to get configs", argMode, argKey))
+		return fmt.Errorf("error getting configs: %w", err)
+	}
+
+	remoteFs, err := cache.Get(context.TODO(), fmt.Sprintf("%s:%s", s.configRcloneRemoteName, s.configPrefix))
+	if err != nil {
+		s.sendMsg(fmt.Sprintf("TRANSFER-FAILURE %s %s failed to get remote fs", argMode, argKey))
+		return err
+	}
+
+	localDir := filepath.Dir(argFile)
+	localFs, err := cache.Get(context.TODO(), localDir)
+	if err != nil {
+		s.sendMsg(fmt.Sprintf("TRANSFER-FAILURE %s %s failed to get local fs", argMode, argKey))
+		return fmt.Errorf("failed to get local fs: %w", err)
+	}
+
+	remoteFileName := argKey
+
+	switch argMode {
+	case "STORE":
+		err = operations.CopyFile(context.TODO(), remoteFs, localFs, remoteFileName, filepath.Base(argFile))
```

Comment:
> This should probably be a two step process: (1) transfer to a temp file with a random component to avoid collisions, (2) rename the temp file to the original key.

Rclone will do this automatically when transferring to a backend where partially uploaded files are visible (eg the local backend or FTP).

> I don't know what happens on every rclone backend if two rename operations race

Depends on the backend but likely last one wins.